### PR TITLE
Feauture/fixing to date from date

### PR DESCRIPTION
**Pull request by nigel-heeral**

### CODE REVIEW

**Comment by nigel-heeral** [comment]
need integration testing

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 1]
```diff
@@ -59,31 +60,32 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
...
+            } catch (ParseException e) {
+                log.error("Error Parsing Date Range for stock:" + stock.getSymbol());
+                return Mono.error(e);
+            }
         }
 
         if (results.size() == 0) {
-            //TODO: Custom exceptions
-            return Mono.error(new Exception("Stock not found"));
+            //TODO: Custom exceptions, B I G Y I K E S
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:72]: explain?

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 2]
```diff
@@ -59,31 +60,32 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
     public Publisher<StockInfo> retrieveStockInfo(StockPayloadDto stockPayload) {
         List<Publisher<StockInfo>> results = new ArrayList<>();
         for (Stock stock : stockPayload.getStocks()) {
-            results.add(retrieveStockInfo(stock.getSymbol(), stockPayload.getFrom(), stockPayload.getTo()));
+            try {
+                results.add(retrieveStockInfo(stock.getSymbol(), stockPayload.getFrom().substring(0,10), stockPayload.getTo().substring(0, 10)));
+            } catch (ParseException e) {
+                log.error("Error Parsing Date Range for stock:" + stock.getSymbol());
+                return Mono.error(e);
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:67]: It's ok for now, but eventually we want different dates for each stock, so throwing error for the whole thing is not feasible

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 3]
```diff
@@ -59,31 +60,32 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
     public Publisher<StockInfo> retrieveStockInfo(StockPayloadDto stockPayload) {
         List<Publisher<StockInfo>> results = new ArrayList<>();
         for (Stock stock : stockPayload.getStocks()) {
-            results.add(retrieveStockInfo(stock.getSymbol(), stockPayload.getFrom(), stockPayload.getTo()));
+            try {
+                results.add(retrieveStockInfo(stock.getSymbol(), stockPayload.getFrom().substring(0,10), stockPayload.getTo().substring(0, 10)));
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:64]: Should we pass the string to the function and let the function handle the validation? Thoughts

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 4]
```diff
@@ -92,53 +94,62 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
...
                     }
                 });
     }
 
-    private StockInfo loadDataFromDb(String ticker, Date fromDate, Date toDate) {
-        List<StockInfo.StockData> data = new ArrayList<>();
-        java.sql.Date start = java.sql.Date.valueOf(fromDate.toInstant().atZone(ZoneId.systemDefault()).toLocalDate());
-        java.sql.Date end = java.sql.Date.valueOf(toDate.toInstant().atZone(ZoneId.systemDefault()).toLocalDate());
+    private boolean isValidDateRange(String fromDate, String toDate) {
+        LocalDate from = LocalDate.parse(fromDate, DATE_FORMAT);
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:114]: So, we dont need to parse it multiple times here

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 5]
```diff
@@ -92,53 +94,62 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
...
+        LocalDate to = LocalDate.parse(toDate, DATE_FORMAT);
+        return from.isBefore(to);
+    }
+
+    private StockInfo loadDataFromDb(String ticker, String fromDate, String toDate) {
+       List<StockInfo.StockData> data = new ArrayList<>();
+        LocalDate start = LocalDate.parse(fromDate, DATE_FORMAT);
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:121]: So, we dont need to parse it multiple times here

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 6]
```diff
@@ -92,53 +94,62 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
...
                 log.warn("Stock {} not found from {} to {}", ticker, start, end);
             }
+
         }, () -> log.warn("Stock {} not found in repository", ticker));
 
         return new StockInfo(data, ticker);
     }
+
+    private static boolean isWeekend(final LocalDate ld) {
+        DayOfWeek day = DayOfWeek.of(ld.get(ChronoField.DAY_OF_WEEK));
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:152]: Holidays still remain to be fixed

File: src/main/java/com/example/tradesimulator/controller/StockService.java [hunk 7]
```diff
@@ -92,53 +94,62 @@ public StockService(StockRepository stockRepository, StockServiceConfig stockSer
...
+            for(StockInfo.StockData stockData : stockList){
+                datesInDb.put(stockData.getDate(), stockData);
+            }
 
-                if (stockDate.equals(end)) {
-                    last = index;
-                } else if (last == -1 && ((end.getTime() - stockDate.getTime()) / (1000 * 60 * 60 * 24)) < 3) {
-                    last = index;
+            boolean storedInDb = true;
+            for (LocalDate date = start; !start.isAfter(end) && !date.isAfter(end); date = date.plusDays(1)) {
```
Comment (rantNYC) [inline src/main/java/com/example/tradesimulator/controller/StockService.java:131]: why do we need !start.isAfter(end), shouldnt start be always before end?

**Review by rantNYC — CHANGES_REQUESTED** [review]
Check my comments

File: src/main/java/com/example/tradesimulator/controller/StockService.java (hunk 1 above)
Comment (nigel-heeral) [reply src/main/java/com/example/tradesimulator/controller/StockService.java:72]: Reee

**Review by rantNYC — APPROVED** [review]

**Pull request merged**